### ISSUE TITLE
Rust backend?

### ISSUE THREAD

**Issue by pmetzger** [skvadrik/re2c#372]
I saw some references in auxiliary materials on re2c.org to a desire to have a rust backend. Has any work been done on this? If not, would help on it be welcome? (And where is the code that handles particular languages located?)

**Comment by skvadrik** [comment]
> Has any work been done on this?

No work so far. I don't know Rust well enough, and I only did a brief investigation to confirm that it is feasible. One of the obstacles is that Rust doesn't have `goto`, but it is easy to overcome by having an integer state variable and a state switch at the beginning of the generated lexer.

> If not, would help on it be welcome?

Definitely yes!

> And where is the code that handles particular languages located?

That would be mostly in https://github.com/skvadrik/re2c/tree/master/src/codegen. Supported languages are [defined here](https://github.com/skvadrik/re2c/blob/master/src/constants.h#L15-L18), so generally you can find most of the language-specific places by searching for text `LANG_`.

- The program is represented as a list of [blocks](https://github.com/skvadrik/re2c/blob/master/src/codegen/code.h#L665-L689), which correspond to `/*!re2c ... */` blocks and directives and the source code in-between. Each block may belong either to the main output file, or to the output header.

- Each block has a number of indented `Code` [elements](https://github.com/skvadrik/re2c/blob/master/src/codegen/code.h#L306-L323), which can be one of the few program constructs defined by the `CodeKind` [enum](https://github.com/skvadrik/re2c/blob/master/src/constants.h#L72-L106).

- The main code generation driver is [here](https://github.com/skvadrik/re2c/blob/master/src/codegen/gen_program.cc#L224-L267).

- The generated code goes through a number of code generation passes, when re2c gradually expands the program control flow graph. Some `Code` elements accumulate whole-program information, so they cannot be generated straight away. Such elements are [expanded later](https://github.com/skvadrik/re2c/blob/master/src/codegen/gen_delayed.cc).

- At last there is a "rendering" pass where most of the language-specific [split is done](https://github.com/skvadrik/re2c/blob/master/src/codegen/render.cc).

- Each language backend has a set of [examples](https://github.com/skvadrik/re2c/tree/master/examples), which should give the idea of the basic functionality that should be covered. 

- Documentation also needs to handle multiple languages (that is done with the help of a [simple script](https://github.com/skvadrik/re2c/blob/master/build/split_man.sh)).

As the previous work on the Go backend has shown, adding a new backend requires a lot of testing with different combinations of existing re2c features. But it is ok to start with something simple and get it working. Some features or special cases may not work out; that is fine, as long as the main cases are working. The Rust work should be done on a branch, as we try to keep `master` in a mostly stable and consistent state. Help is welcome, including any insights or examples.

**Comment by pmetzger** [comment]
FYI, it appears LLVM has just added a hack that optimizes DFAs created with loop-around-switch style constructs. See https://reviews.llvm.org/rG02077da7e7a8 which will presumably be in a release in a few months. So Rust won't suffer too much for the lack of goto statements.

**Comment by skvadrik** [comment]
Jump threading is useful. But even without it I don't expect the performance to be worse than a hand-written or a table-based lexer in Rust. That is only my intuition, of course.

> I don't recall seeing an option in the manual for constructing a loop with embedded switch style of code generation, so I suppose that would need to be done afresh. However, I also presume that could optionally be used by both the C and Go backends as well, so one way to test that particular thing would be to generate C code of that form and then make sure all the tests still pass?

Right, re2c doesn't have an option for that. I don't think it would generate faster code for C and Go backends, but it would be interesting to try, if only to estimate the difference between this and the usual `goto`-based lexer. It is a good idea to implement it for C first.

And sorry for a late reply!

**Comment by pmetzger** [comment]
Even ignoring the performance estimation (which would be interesting), it seems like it would be easier to build the loop-and-switch code for Rust if it was already built for C. (Also of interest at some point might be a tail recursive version; several C compilers now have annotations to force tail recursion. Unfortunately the Rust compiler does not yet have tail recursion optimization.)

**Comment by Raekye** [comment]
> `switch` statements always have an explicit `goto`, and `if`/`else` statements may omit `goto` if the branch target happens to be the next state, but this logic can be turned off (the relevant code is [here](https://github.com/skvadrik/re2c/blob/master/src/codegen/code_goto.cc#L136-L154), the `== next` conditions).
> 
> Can you give an example of what would be a problem? I'm not sure I understand correctly.

I just double checked the generated lexer for my project, and it has code like:

```
yyc_newline:
        if (((self->buffer + self->length + YYMAXFILL) - (self->cursor)) < 2) yyfill(2);
        yych = *(self->cursor);
        switch (yych) {
        case '\t':      goto yy4;
        case '\n':      goto yy7;
        case '\r':      goto yy9;
        case ' ':       goto yy10;
        case '#':       goto yy13;
        default:        goto yy2;
        }
yy2:
        ++(self->cursor);
yy3:
// more code
```

Hopefully I'm not missing something stupid - but the label/state `yy2` depends on falling through to `yy3` correct? Off the top of my head, translating this to a loop-switch in `C` seems pretty straightforward since a switch really just contains labels with the same fall-through semantics. To make sure I'm on the same page, it would look something like this:

```
enum label_t {
    yyc_newline,
    yyc_2,
    yyc_3,
   // ...
};

label_t current_label = ...;
while 1 {
    switch current_label {
    case yyc_newline:
        if (((self->buffer + self->length + YYMAXFILL) - (self->cursor)) < 2) yyfill(2);
            yych = *(self->cursor);
            switch (yych) {
            case '\t':      current_label = yy4; continue;
            case '\n':      current_label = yy7; continue;
            case '\r':      current_label = yy9; continue;
            case ' ':       current_label = yy10; continue;
            case '#':       current_label = yy13; continue;
            default:        current_label = yy2; continue;
            }
        }
    case yy2:
        ++(self->cursor);
        // this still falls through to yy3 as in the original code
    case yy3:
        // more code
    }
}
```

Basically, I would turn `goto xxx;` into `current_label = xxx; continue;`.

But in Rust, besides not having gotos, its `match` construct (the closest thing to a C switch) does not fall-through. I only took a quick look at [code_goto.cc](https://github.com/skvadrik/re2c/blob/master/src/codegen/code_goto.cc#L136-L154); I'm not familiar with exactly where/how it's used, but are you saying by disabling the ` == next`, label `yy2` would get an explicit `goto yy3`?

I'm far from a Rust expert, but I have been coding in it as my primary language for about 6 months (I would say it's not particularly hard to pick up once you're forced to use it) - I don't think we need anything too fancy so hopefully I can answer any basic Rust questions!

**Comment by skvadrik** [comment]
Ah, I see. You can avoid fallthrough like this:
```
while 1 {
 switch current_label {
    // ...
    case yy2:
        ++(self->cursor);
        current_label = yy3;
        continue;
    case yy3:
        // more code
    }
```
Essentially fallthrough from `yy2` to `yy3` is like a switch where all cases have a `goto yy3`, so they are omitted for brevity. But you can bring the `goto yy3` back and it will not change anything. In Rust it will be more work because we have to reiterate on the state variable.

> hopefully I can answer any basic Rust questions!

Great!

**Comment by Raekye** [comment]
> Essentially fallthrough from `yy2` to `yy3` is like a switch where all cases have a `goto yy3`, so they are omitted for brevity. But you can bring the `goto yy3` back and it will not change anything. In Rust it will be more work because we have to reiterate on the state variable.

To be sure, this behaviour is exactly controlled by the `== next` in [code_goto.cc](https://github.com/skvadrik/re2c/blob/master/src/codegen/code_goto.cc#L136-L154)? Again forgive me I'm not too familiar with the code - it seems that file is language-agnostic. Naively, I would want to just guard `... == next` with `&& opts->lang != LANG_RUST` (of course, we could just propagate that information into this function, but then it would nolonger be language-agnostic).

> > I'm not sure that would really be of much help in practice
> 
> It would, because my knowledge of Rust is lacking and having an example makes it easier. That's how the Go backend started: #272.

Does this mean you will be attempting a Rust backend, given examples? I.e. would you recommend attempting a Rust backend on my own, or would it be best to just provide more examples?

**Comment by Raekye** [comment]
I started looking at implementing things since I was excited 😅 

> The program is represented as a list of [blocks](https://github.com/skvadrik/re2c/blob/master/src/codegen/code.h#L665-L689), which correspond to `/*!re2c ... */` blocks and directives and the source code in-between. Each block may belong either to the main output file, or to the output header.

From my understanding, we need to wrap each `/*!re2c ... */` block with (in C):

```
// initializers
while (1) {
    switch (current_label) {
    case ...:
        // code following the label
        // goto xxx;
        // becomes
        current_label = xxx;
        continue;
    }
}
```

In Rust, it would look something like this:

```
// initializers
'yyloop: loop {
    match current_label {
        label1 => {
            // body
            current_label = xxx;
            continue 'yyloop;
        },
        label2 => {
            // body
            current_label = xxx;
            continue 'yyloop;
        },
    }
}
```

In particular, the the code after each label needs to be wrapped in a block; each "label"/state has an associated "body". However, it's not so obvious to me that there's a distinction in re2c currently - as in, the code following a label is simply outputted without concern for the label it's under. I'm not sure if I'm making myself clear. But if so, how would you recommend approaching this?

Also, it seems to me that conceptually each start condition generates a DFA? In [code.h](https://github.com/skvadrik/re2c/blob/master/src/codegen/code.h#L294-L304), is it true that `NLABEL` corresponds to a label within a DFA/start condition, whereas `SLABEL` corresponds to the different DFAs/start conditions?

**Comment by skvadrik** [comment]
>To be sure, this behaviour is exactly controlled by the == next in code_goto.cc?

Yes. You can pretend that `== next` condition is never satisfied by commenting out the cases that have it (temporarily). For Rust this particular function will not be used at all, since it generates `if`/`else` statements and we will probably not need them.

> Does this mean you will be attempting a Rust backend, given examples? I.e. would you recommend attempting a Rust backend on my own, or would it be best to just provide more examples?

If I have some examples to start with, I can work on the Rust backend (this is something I wanted to do for a while, but kept delaying). But if you want to do it yourself, I can surely help. It's not easy --- you'll need to learn the codegen subsytem in re2c to understand its architecture, as it will most likely need some restructuring. It is also completely fine if you try and abandon it later, I can resume the work.

> I started looking at implementing things since I was excited

\o/

The simplest thing to start with (as @pmetzger suggested) is to change the C backend to use the loop-switch approach that we will need for Rust. This is not necessary, but it's a good exercise and it will allow you to familiarize with the codebase and see if you want to work on it at all. You don't have to make loop-switch work for all the cases, just hack existing code enough to make some simple examples work.

The first thing for a Rust backend is still to have a few complete examples of lexers in Rust, just to know what code we will need to generate. Otherwise it may turn out later that we forgot about some crucial language feature that blocks all the previous work and requires major restructuring of the code. 

> In particular, the the code after each label needs to be wrapped in a block; each "label"/state has an associated "body". However, it's not so obvious to me that there's a distinction in re2c currently - as in, the code following a label is simply outputted without concern for the label it's under. I'm not sure if I'm making myself clear. But if so, how would you recommend approaching this?

There is a `CodeBlock` construct: https://github.com/skvadrik/re2c/blob/master/src/codegen/code.h#L248-L257. It can have different formatting options, including `WRAPPED` which is rendered as indented block wrapped in curly braces (this is what you need). For examples of usage see https://github.com/skvadrik/re2c/search?q=WRAPPED.

> Also, it seems to me that conceptually each start condition generates a DFA? 

Correct. The loop https://github.com/skvadrik/re2c/blob/master/src/codegen/gen_dfa.cc#L142 handles it. I recommend that you don't worry about conditions (`-c` flag) now, this is a more complex mode that can be handled later.

> In code.h, is it true that NLABEL corresponds to a label within a DFA/start condition, whereas SLABEL corresponds to the different DFAs/start conditions?

As it is implemented now, yes: `NLABEL` is a DFA state label like `yy3`, `yy12345` and `SLABEL` is some special label like a custom user-defined start label. More generally `NLABEL` is a numeric label, and `SLABEL` is a string label.

**Comment by Raekye** [comment]
> > To be sure, this behaviour is exactly controlled by the == next in code_goto.cc?
> 
> Yes. You can pretend that `== next` condition is never satisfied by commenting out the cases that have it (temporarily). For Rust this particular function will not be used at all, since it generates `if`/`else` statements and we will probably not need them.

I noticed that code uses [`add_branch` in `code_goto.cc`](https://github.com/skvadrik/re2c/blob/master/src/codegen/code_goto.cc#L124) which sets an `elide` field. Then [`gen_goto` in `gen_state.cc`](https://github.com/skvadrik/re2c/blob/master/src/codegen/gen_state.cc#L476-L485) checks it. Would it work to just update `gen_goto` to never elide if the "loop-switch" flag is on? E.g. `gen_state.cc` also has [`gen_fill_falllback`](https://github.com/skvadrik/re2c/blob/master/src/codegen/gen_state.cc#L327-L340) which has a comment about eliding; I'm guessing this fall-through needs to be explicit if loop-switch is used. Or would you still recommend changing `add_branch` instead (updating `== next` to also check `!opts->loop_switch`)? (It would seem `gen_fill_falllback` would still need to be handled separately - are there any other such cases?)

> If I have some examples to start with, I can work on the Rust backend (this is something I wanted to do for a while, but kept delaying). But if you want to do it yourself, I can surely help. It's not easy --- you'll need to learn the codegen subsytem in re2c to understand its architecture, as it will most likely need some restructuring. It is also completely fine if you try and abandon it later, I can resume the work.

I'm having fun so far messing around with it (adding a loop-switch backend for C/Go), but if @pmetzger can provide you with some examples this weekend (or whenever they have time), I wouldn't hesitate to start working on it yourself just because I've been messing around (assuming you also have time). Of course, at that time we can see if I have any meaningful progress to share

---
I joined the re2c IRC channel on Libera, in case it's easier to continue discussions there. Of course, I don't want to take up too much of your weekend :P So if it's better to keep it in this github issue that's perfectly fine with me

**Comment by skvadrik** [comment]
Let's move the discussion to IRC. I'll reply about the code there.

**Comment by pmetzger** [comment]
I'll try to join the re2c channel tomorrow, I didn't realize it existed.

BTW, @Raekye, as what you're creating is a state machine, it's conventional to call the state variable something like "state" rather than "current_label", but other than that, what you suggested for the Rust code looks approximately correct. You would almost certainly want to make "state" a rust enum type and generate an enum for every label.

(Equivalent code can be generated in C of course.)

A somewhat bigger problem is figuring out the necessary data structures for handling the input. re2c lexers want to be able to do things like keeping cursors into the existing input and the like. I imagine whatever is done in Go might be a good model; Rust doesn't allow use of raw pointers by default, but Go is also safe in this respect, and has similar slice types.

**Comment by pmetzger** [comment]
By "data structures for handling input", I mean, in the C API we have things like YYCURSOR,  YLIMIT, YYPEEK, etc. etc., and one needs to handle these reasonably in Rust. Presumably in Rust one would usually use a vector of YYCTYPE as an input buffer, have a YYFILL that refills the buffer, etc. The spirit of re2c would be to allow great flexibility in the implementation of these; luckily Rust has inline functions, and even macros, so it should still be possible to hit the limits of what hand coded lexers can do.

**Comment by skvadrik** [comment]
> A somewhat bigger problem is figuring out the necessary data structures for handling the input. re2c lexers want to be able to do things like keeping cursors into the existing input and the like. I imagine whatever is done in Go might be a good model; Rust doesn't allow use of raw pointers by default, but Go is also safe in this respect, and has similar slice types.

Right. Figuring out how to adapt re2c input model to Rust is the real challenge here. We can use byte buffers and indices instead of pointers (as it is with Go), but we should explore all options and find fastest one. Wrapping the generated code in "unsafe" is fine if makes it faster. That's why we need complete working examples of lexers in Rust before attempting an implementation (preferably from someone who knows Rust and can validate that the code looks reasonable).

The loop-switch feature for C is a good starter exercise and it is of some academic interest, but we shouldn't spend too much time on it --- without the Rust backend it will be additional implementation complexity without practical value. Figuring out the interaction between Rust and re2c is more important at the moment.

**Comment by pmetzger** [comment]
`unsafe` is almost certainly not needed. The only two places it would come up would be if we were using raw pointers (unnecessary for performance, the compiler deals with array indexing quite efficiently) and if we were coercing a byte array into a utf8 string without checking for validity (we might want this, but if we did, the standard library already provides a routine for it which the compiler will inline into a noop.)

**Comment by pmetzger** [comment]
Per @skvadrik I'm noting here one instance where `unsafe` might be useful; we may want a an option to re2rust to do unsafe array indexes to avoid bounds checks for performance reasons in some lexers.

**Comment by skvadrik** [comment]
Update: the first working example for the Rust backend has been merged into master: https://github.com/skvadrik/re2c/commit/0c0d674e1dcdd54826ff8ab863c8c1cf6bcce113. Commits https://github.com/skvadrik/re2c/commit/01046e779caca410f6769ba4be94efb765963198 and https://github.com/skvadrik/re2c/commit/3c02670f0f0a228b7122fc0d0e8b1a5a8619f94e added `re2rust` tool for Autotools and CMake build systems. It's a start! Special thanks for @Raekye and @pmetzger for getting the ball rolling.

**Comment by pmetzger** [comment]
Whoa!

**Comment by pmetzger** [comment]
@skvadrik Do you need me to start porting more examples? (Sorry I've been scarce.)

**Comment by skvadrik** [comment]
@pmetzger No worries, I can go on myself for now. It almost invariably requires fixing some bit of the code generator to make each example work (so you'd be blocked). But it would be good if you take a look at the examples when they are ready and let me know if they can be made more idiomatic. By "ready" I mean when I stop changing them, which I'm doing frequently  now as I learn new bits of Rust.

**Comment by pmetzger** [comment]
@skvadrik I'm going to check out the repo and I'll get comments to you.

**Comment by pmetzger** [comment]
I note you're generating `unsafe` blocks but I don't see any unsafe code in them other than `get_unchecked` and the like. Do you actually need `unsafe`?

**Comment by skvadrik** [comment]
I do need unsafe for `get_unchecked`, otherwise I'm getting compilation error:
```
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
```
I thought of providing an option `--no-unsafe` in case the user wants to debug the generated code (or they don't want `unsafe` for some other reason), but the default should allow check-less buffer indexing, because the generated lexer does it already (and not necessarily in the form of bounds checks).

**Comment by pmetzger** [comment]
Ah, I understand. The thing is, you probably want to minimize the size of the unsafe block. I'll ponder it.

**Comment by pmetzger** [comment]
Btw, thinking about it for literally only a couple of minutes: the user should be able to provide the `get` function so they can pick whether it's a `get` or `get_unsafe`, or should at least be able to tell `re2rust` which to use. Especially when debugging one is going to want to use `get`.

**Comment by pmetzger** [comment]
I've realized we have a small problem that didn't occur to me earlier. Rust documentation comments start with `/*!` or `//!`

**Comment by skvadrik** [comment]
> you probably want to minimize the size of the unsafe block

Now `unsafe` is only around `YYPEEK` (https://github.com/skvadrik/re2c/commit/62c3c6a5674d3059a211901ba8465bf3f01144d3).

> the user should be able to provide the get function so they can pick whether it's a get or get_unsafe, or should at least be able to tell re2rust which to use

Yes, I'm thinking about it as well. I personally use the bounds-checked version to validate examples, and it's immensely useful. First thing that comes to mind is: the user implements `YYPEEK` as a macro depending on debug/release mode (hopefully Rust allows that, but I haven't tried) and re2c provides an option `--no-unsafe` that drops the `unsafe` wrappers around `YYPEEK` (to avoid all the warnings about useless `unsafe`). Since re2rust does not generate `YYPEEK` definitions, it's impossible to fully automate this without the need to provide the macro on the user side.

> I've realized we have a small problem that didn't occur to me earlier. Some Rust documentation comments start with `/*!` or `//!`. 

Do you think people would run the doc tool on .re files? If they do, what will happen to the blocks --- will they result in unwanted code blocks in documentation, or will the doc tool emit an error? it is feasible to use other begin/end markers, but I'd rather avoid it if possible.

**Comment by pmetzger** [comment]
> Do you think people would run the doc tool on .re files? 

I suppose one would have to prohibit that. It might be okay provided one warns about the issue in the documentation?

**Comment by pmetzger** [comment]
Oh, and on the `get` (or `[]`) vs. `get_unsafe` issue, I think provided the user can pick (which they can) and provided the generator won't produce `unsafe`s when it's not needed (thanks to the `--no-unsafe` flag), we're fine.

**Comment by skvadrik** [comment]
Rust is supported now (all examples have been implemented). I am updating the docs, and I plan to release re2c-3.0 after I do some pre-release testing and finish the documentation work.

**Comment by pmetzger** [comment]
I'm already using it in two projects, but having an actual release will be spectacular! Are you using the C lexer example I contributed?

**Comment by skvadrik** [comment]
No, I thought you were still improving it. Can you attach the latest version?

**Comment by pmetzger** [comment]
The last I sent to you is good enough for now. I may send PRs with improvements over time.

**Comment by pmetzger** [comment]
(Let me know if you can't find it and I'll email it again.)

**Comment by skvadrik** [comment]
No, I have, thanks. I will add it to the examples and let you know.

**Comment by pmetzger** [comment]
Oh, and just for avoidance of doubt, as with the rest of re2c, I disclaim all copyright; that code is in the public domain.

**Comment by skvadrik** [comment]
Here it goes: https://github.com/skvadrik/re2c/commit/38a07a33043dbe20d0fb06484122ba09961731fa

**Comment by pmetzger** [comment]
Cool! BTW, did you add that macro_rules! log thing? I don't remember adding it and it doesn't seem to be used in the file.

**Comment by skvadrik** [comment]
Yes, I changed a few things to fit it into the re2c testing framework (it expects certain things, e.g. that by default the code runs without any messages and returns either zero on success or non-zero on failure). I did not change anything in the example itself (re2c rules, lexer interactions, etc.).

**Comment by skvadrik** [comment]
Finally this can be closed. \o/

**Comment by pmetzger** [comment]
Now we need to advertise this on reddit's r/rust when the release happens.

**Comment by skvadrik** [comment]
That would be great! I will write release notes as usual, they can be used as a reference.

**Comment by pmetzger** [comment]
I will flog it heavily. In my opinion, no other lexical analyzer generator comes close to re2.+ on any axis; it is the most flexible to work with and produces the fastest lexers with the least effort on the part of the programmer.

**Comment by pmetzger** [comment]
See:
https://www.reddit.com/r/rust/comments/se4khc/re2c_lexical_analyzer_generator_now_supports_rust/

**Comment by skvadrik** [comment]
Ho ho!! Super nice, thank you! :D

**Comment by pmetzger** [comment]
It's such a good tool, I feel like everyone should know about it!
